### ISSUE TITLE
Mac M1 Second Generation: Renders are black

### ISSUE THREAD

**Issue by gkjohnson** [gkjohnson/three-gpu-pathtracer#180]
Gradients show in the main render demo but the geometry itself is black. Possibly the environment maps are not loading correctly?

**Comment by bhouston** [comment]
It is happen on my Mac mini M1 that is on MacOS 16 beta.  But only in Safari.

Works in Chrome:

![Screen Shot 2022-06-13 at 1 46 52 PM](https://user-images.githubusercontent.com/588541/173414236-94dc2463-233d-46e8-bb5a-285490bcba53.png)

Doesn't work in Safari:

![Screen Shot 2022-06-13 at 1 46 06 PM](https://user-images.githubusercontent.com/588541/173414234-f08c2e52-bc56-46f9-aed5-718cfbef9568.png)
![Screen Shot 2022-06-13 at 1 48 49 PM](https://user-images.githubusercontent.com/588541/173414242-107de3cf-dee5-46f5-8e56-6cd560197021.png)
![Screen Shot 2022-06-13 at 1 50 59 PM](https://user-images.githubusercontent.com/588541/173414243-403897a6-99be-4a25-99dd-55fdc54feddf.png)

Notice the gelatinous cube shows up but every other material doesn't.  I tried to change all settings in the example viewer and nothing makes a difference.  The WebGL three.js preview does work, but soon as it switches to path tracing it breaks.

**Comment by bhouston** [comment]
Most scenes render black but a few that seem to have emissive do seem to work slightly:


![Screen Shot 2022-06-13 at 1 53 13 PM](https://user-images.githubusercontent.com/588541/173414972-d86d443f-eda6-438b-9cf5-2ee02499e327.png)
![Screen Shot 2022-06-13 at 1 54 00 PM](https://user-images.githubusercontent.com/588541/173414982-ed414d87-b0de-40da-b28d-ddb1d2bbaef3.png)
![Screen Shot 2022-06-13 at 1 54 23 PM](https://user-images.githubusercontent.com/588541/173414984-d5e6ce7e-4825-4fb2-a1ac-9637a06e6ac9.png)

**Comment by gkjohnson** [comment]
Seems like a texture issue, then 🤔

**Comment by bhouston** [comment]
@gkjohnson did you manage to get a Mac mini M1 yet?  It would reproduce it.

**Comment by gkjohnson** [comment]
> did you manage to get a Mac mini M1 yet? It would reproduce it.

I did! That's how I found this. 

Looks like textures are working okay but there's possibly an issue with blending, at least? And it seems the shader compiler is sensitive variable declaration order relating to random seed initialization... moving the seed declaration up in the list of variables lets the random noise update.

Here's an image early-outing the bg color and albedo showing that textures are definitely being sampled.

<img width="670" alt="image" src="https://user-images.githubusercontent.com/734200/173449031-e5b41811-2bdc-4809-9c1e-dbd0250a40b5.png">

**Comment by gkjohnson** [comment]
Renderer seems to have been working on May 17th:

https://twitter.com/BenHouston3D/status/1526560267957592065

A working commit:

3350a07176d53fdf10e7cb6a9fee69ba5cb8fb2e

**Comment by setpixel** [comment]
Also, the rendering hangs on M1 ipad in safari and chrome. I don't know if it's related.

**Comment by setpixel** [comment]
Actually, I just checked on my wife's non m1 iPhone. Same result, rendering locks. I wonder if the iOS issue is a different issue entirely? Should I make a new / separate issue?

**Comment by bhouston** [comment]
> Also, the rendering hangs on M1 ipad in safari and chrome. I don't know if it's related.

On iPadOS and iOS all browser (e.g. Firefox, Chrome) are actually Safari with different skins.   But on MacOS, Chrome is actually Chrome.  This is very confusing but it is the current situation.

Thus it is still likely a Safari only bug and not a hardware issue.

**Comment by bhouston** [comment]
More details on why Chrome, Firefox is actually Safari on iOS/iPadOS: https://news.ycombinator.com/item?id=25850091

**Comment by gkjohnson** [comment]
It looks like this broke with #171 (tested commit 1334237) while it worked after #165 (tested commit 08b1315)

> Actually, I just checked on my wife's non m1 iPhone. Same result, rendering locks. I wonder if the iOS issue is a different issue entirely? Should I make a new / separate issue?

Thank! Let me track down the issue first and if it doesn't fix the issue on iOS we can track that down, too.

**Comment by gkjohnson** [comment]
Okay so I was looking at the PRs out of order - it was actually #162 that broke Safari with orthographic camera support... Changing [this section of code](https://github.com/gkjohnson/three-gpu-pathtracer/blob/8d76851d98e7bee9a510e74b252ef3a396fceffc/src/materials/PhysicalPathTracingMaterial.js#L326-L336) from this:

```js
if ( isOrthographicCamera ) {
    rayDirection = ( cameraWorldMatrix * vec4( 0.0, 0.0, -1.0, 0.0 ) ).xyz;
    rayDirection = normalize( rayDirection );
} else {
    vec3 cameraOrigin = ( cameraWorldMatrix * vec4( 0.0, 0.0, 0.0, 1.0 ) ).xyz;
    rayDirection = normalize( rayOrigin - cameraOrigin );
}
```

to this:

```js
// if ( isOrthographicCamera ) {
//    rayDirection = ( cameraWorldMatrix * vec4( 0.0, 0.0, -1.0, 0.0 ) ).xyz;
//    rayDirection = normalize( rayDirection );
// } else {
    vec3 cameraOrigin = ( cameraWorldMatrix * vec4( 0.0, 0.0, 0.0, 1.0 ) ).xyz;
    rayDirection = normalize( rayOrigin - cameraOrigin );
// }
```

ie removing the `isOrthographicCamera` condition - fixes the problem. I wonder if boolean uniforms are somehow broken in Safari?

**Comment by gkjohnson** [comment]
Nope even with changing it to a float or int it's still broken. I think changing to a `#define` is our best option here.

**Comment by gkjohnson** [comment]
I should have fixed this in #187 - @bhouston @setpixel can you guys give this a try, now?

https://gkjohnson.github.io/three-gpu-pathtracer/example/bundle/materialBall.html

![image](https://user-images.githubusercontent.com/734200/174401576-d772ca32-ed07-40ab-b27a-a551815d5083.png)

**Comment by setpixel** [comment]
Yes! That works. However, the interior scene, the depth of field, skinned mesh, morph, ambient occlusion demos do not render.

**Comment by gkjohnson** [comment]
Hmm - those all work for me... Do you mind making a new issue with the device details that's not working? And if you can any logs / errors that the device is logging?

**Comment by setpixel** [comment]
Hmm! Are you sure you don't get a page crash on https://gkjohnson.github.io/three-gpu-pathtracer/example/bundle/depthOfField.html in Safari? I can't seem to debug anything, the tab just crashes on M1 Macbook Air Safari. 

I'll create a new issue.

**Comment by gkjohnson** [comment]
Yeah it's definitely working - is there any value in clearing the cache? Does it work on your iphone?

<img width="1163" alt="image" src="https://user-images.githubusercontent.com/734200/174460035-ea8a2801-0d06-4ce9-9368-ae81af4f2833.png">
